packages/ozone/src/api/moderation/scheduleAction.ts

```diff
@@ -0,0 +1,118 @@
+import { ToolsOzoneModerationScheduleAction } from '@atproto/api'
+import { AuthRequiredError, InvalidRequestError } from '@atproto/xrpc-server'
+import { AppContext } from '../../context'
+import { Server } from '../../lexicon'
+import { subjectFromInput } from '../../mod-service/subject'
+import { ExecutionSchedule } from '../../scheduled-action/types'
+import { getScheduledActionType } from '../util'
+
+export default function (server: Server, ctx: AppContext) {
+  server.tools.ozone.moderation.scheduleAction({
+    auth: ctx.authVerifier.modOrAdminToken,
+    handler: async ({ input, auth }) => {
+      const access = auth.credentials
+      const db = ctx.db
+      const { action, subjects, createdBy, scheduling, modTool } = input.body
+
+      if (!access.isModerator) {
+        throw new AuthRequiredError('Must be a moderator to schedule actions')
+      }
+
+      if (access.type === 'admin_token' && !createdBy) {
+        throw new AuthRequiredError(
+          'Must specify createdBy when using admin auth',
+        )
+      }
+
+      const actionType = getScheduledActionType(
+        action.$type?.split('#')[1] || '',
+      )
+
+      const succeeded: string[] = []
+      const failed: { subject: string; error: string; errorCode?: string }[] =
+        []
+
+      // Defining alternatively required fields is not supported by lexicons so we need to manually validate here
+      if (!scheduling.executeAt && !scheduling.executeAfter) {
+        throw new InvalidRequestError('Must specify an execution schedule')
+      }
+
+      const executionSchedule: ExecutionSchedule = scheduling.executeAt
+        ? { executeAt: new Date(scheduling.executeAt) }
+        : {
+            executeAfter: new Date(scheduling.executeAfter!),
+            executeUntil: scheduling.executeUntil
+              ? new Date(scheduling.executeUntil)
+              : undefined,
+          }
+
+      const eventData = { ...action, modTool }
+      const actualCreatedBy =
+        access.type === 'admin_token'
+          ? createdBy || ctx.cfg.service.did
```

Comment:
I think `|| ctx.cfg.service.did` is a dead codepath due to the validation above.  Maybe just worth noting/reflecting that if it's not easy to eliminate the codepath.